boto3utils/s3.py

```diff
@@ -116,36 +122,88 @@ def upload_json(self, data, url, extra={}, **kwargs):
         finally:
             rmtree(tmpdir)
 
-    def get_object(self, bucket, key):
+    def get_object(self, bucket, key, extra_args={}):
         """ Get an S3 object """
+        extra_args = deepcopy(extra_args or {})
+
         if self.requester_pays:
-            response = self.s3.get_object(Bucket=bucket,
-                                          Key=key,
-                                          RequestPayer='requester')
-        else:
-            response = self.s3.get_object(Bucket=bucket, Key=key)
-        return response
+            extra_args["RequestPayer"] = "requester"
 
-    def download(self, uri, path=''):
+        return self.s3.get_object(Bucket=bucket, Key=key, **extra_args)
+
+    def download(self, uri, path='', extra_args={}):
         """
         Download object from S3
         :param uri: URI of object to download
         :param path: Output path
+        :param extra_args: Dictionary of parameters to pass through to the S3 client
         """
         s3_uri = self.urlparse(uri)
         fout = op.join(path, s3_uri['filename'])
         logger.debug("Downloading %s as %s" % (uri, fout))
-        if path != '':
+
+        if path != "":
             makedirs(path, exist_ok=True)
-        extra_args = None
+
+        extra_args = deepcopy(extra_args)
         if self.requester_pays:
-            extra_args = {'RequestPayer': 'requester'}
-        self.s3.download_file(s3_uri['bucket'],
-                              s3_uri['key'],
+            extra_args["RequestPayer"] = "requester"
+
+        if s3_uri["query_params"]:
+            for key in s3_uri["query_params"]:
+                extra_args[key] = s3_uri["query_params"][key]
+
+        self.s3.download_file(s3_uri["bucket"],
+                              s3_uri["key"],
                               fout,
                               ExtraArgs=extra_args)
         return fout
 
+    def download_with_metadata(self,
```

Comment:
I'm not super fond of `download` and `download_with_metadata` having two different ways (boto3 download_fiole vs get_object), but it was really due to how this library before had `download` use the `download_file` function rather than `get_object`, which is really the more useful function, clearly - since it provided back the metadata as well. 

I'm not sure of the performance implication....is `download_file` any faster?  Because if not, then `download` should really be also changed to call `download_with_metadata` and simply return the filename. This will keep backward compatibility and reduce duplication of code.

Reply:
There is actually a difference. I'm not sure the best approach to take here:

* `boto3.s3.download_file` is a managed transfer service. This means for large files it performs multiple range-gets to download the file in parallel. This can lead to significant performance gains.
* `boto3.s3.get_object` provides the file as a `StreamingBody` which can then be written to disk, or processed directly in memory. But you get the full object data, not just the file itself.

There are several options i can think of here (and probably many more that aren't coming to mind):
1. leave the 2 methods as they are, with different implementations
2. add a `get_file_metadata` method that passes through to the  `boto3.s3.get_object_attributes` method to download the file's metadata. Then update the `download_file_with_metadata` method to call both `get_file_metadata` and the existing `download_file`
3. your reco above, update the `download_file` to simply use the `get_object` under the covers.

Each of these have their own pros/cons... i'm not sure which is better.  i maybe lean toward number 2, but that means every call to `download_file_with_metadata` will result in 2 S3 calls, not just one. ... maybe that's ok?